### PR TITLE
module: add background and border colors

### PR DESCRIPTION
**Pull request by lasers**
This only works on `i3-gaps`. I add `background` and `border` colors as well as `border_{top, bottom, left, right}` widths. The border widths only work if you specified `border`.

2018-07-23-112512_3840x1200_scrot
2018-07-23-120345_3840x1200_scrot_
2018-07-23-121350_3840x1200_scrot
2018-07-23-154623_3840x1200_scrot

### CODE REVIEW

File: py3status/module.py [hunk 1]
```diff
@@ -350,6 +351,29 @@ def set_module_options(self, module):
...
 
+        for name in ('background', 'border'):
+            if name in mod_config:
+                param = mod_config[name]
+                if not isinstance(param, basestring):
+                    err = 'invalid "{}" attribute, '.format(name)
+                    err += 'should be a string'
+                    raise TypeError(err)
+                self.py3_module_options[name] = Py3(self)._get_color(param)
+                self.block_options.append(name)
```
Comment (lasers) [inline py3status/module.py]: @tobes Look at this part. I don't know if `Py3(self)._get_color(param)` is ideal.

Comment (tobes) [reply py3status/module.py]: yes this is very bad

a) Instantiating `Py3` just to access a function is a performance overhead

b) class methods starting with an underscore `_` are private and should not be accessed outside of the class

I'll make a PR to make the function globally available

**Comment by lasers** [comment]
This is right now. Something changed. I'm not sure what's going on. :-)

2018-09-02-112751_727x133_scrot

**Comment by lasers** [comment]
@tobes Fixed. Ready for review. I included `param` in the errors too.

File: py3status/module.py [hunk 2]
```diff
@@ -350,6 +350,35 @@ def set_module_options(self, module):
...
             self.i3s_module_options['separator_block_width'] = sep_block_width
 
+        for name in ('background', 'border'):
+            if name in mod_config:
+                param = mod_config[name]
+                if not isinstance(param, basestring):
+                    raise TypeError(
+                        'Invalid attribute `{}`, '
+                        'should be an int, not `{}`'
+                        ''.format(name, param)
```
Comment (tobes) [inline py3status/module.py]: you can remove the `''`

```python
(
    'I like parenthesis {} much, '
    'they let me have such fun.'
    .format('so')
)
```

The wording is confusing it sounds like you are going to give the type, either specify the type or better just say `got ...` or similar

File: py3status/module.py [hunk 3]
```diff
@@ -380,6 +409,7 @@ def process_composite(self, response):
         # update all components
         color = response.get('color')
         urgent = response.get('urgent')
+        length = len(response['composite']) - 1
```
Comment (tobes) [inline py3status/module.py]: `composite_length` for clarity, but nice moving it out of the loop

File: py3status/module.py [hunk 4]
```diff
@@ -412,6 +442,14 @@ def process_composite(self, response):
             elif urgent and 'urgent' not in item:
                 item['urgent'] = urgent
 
+            # set background and border colors
+            for key, value in self.py3_module_options['block'].items():
+                if (key == 'border_left' and index != 0) or (
+                        key == 'border_right' and index != length):
```
Comment (tobes) [inline py3status/module.py]: what are we doing here a comment is always nice.

something like `we only want side borders at start/end of a composite`

File: py3status/module.py [hunk 5]
```diff
@@ -311,7 +311,7 @@ def set_module_options(self, module):
         https://i3wm.org/i3status/manpage.html#_universal_module_options
         """
         self.i3s_module_options = {}
-        self.py3_module_options = {}
+        self.py3_module_options = {'block': {}}
```
Comment (tobes) [inline py3status/module.py]: this is ugly, I think we can do better

File: py3status/module.py [hunk 6]
```diff
@@ -325,7 +325,7 @@ def set_module_options(self, module):
 
             if 'align' in mod_config:
                 align = mod_config['align']
-                if not (isinstance(align, basestring) and
+                if not isinstance(align, basestring) and (
```
Comment (tobes) [inline py3status/module.py]: why have you done this?
explain your working and why is this not in the PR description

File: py3status/module.py [hunk 7]
```diff
@@ -350,6 +350,35 @@ def set_module_options(self, module):
...
+                if not isinstance(param, basestring):
+                    raise TypeError(
+                        'Invalid attribute `{}`, '
+                        'should be an int, not `{}`'
+                        ''.format(name, param)
+                    )
+
+                self.py3_module_options['block'][name] = (
+                    Py3(self)._get_color(param)
+                )
```
Comment (tobes) [inline py3status/module.py]: what's going on here? you need to have comments explaining stuff.  This is changes to core py3status they need to be maintainable.

Any PR should leave the code nicer than before :)

File: py3status/module.py (hunk 6 above)
Comment (lasers) [reply py3status/module.py]: Travis.

File: py3status/module.py (hunk 5 above)
Comment (lasers) [reply py3status/module.py]: Used to be `self.block_option`. It's not going to be used for anything else other than `background`, `border` + its widths. Happy to hear your suggestion.

File: py3status/module.py (hunk 7 above)
Comment (lasers) [reply py3status/module.py]: This is same as before, except that I'm putting everything inside `self.py3_module_options = {'block': {}}` instead of (2) separate parameters.

File: py3status/module.py (hunk 6 above)
Comment (tobes) [reply py3status/module.py]: I don't see why it is part of this PR maybe I've missed something 
and please don't resolve issue I raise I'm not happy

Also I don't have to help you with your code I can just NAK it

File: py3status/module.py (hunk 1 above)
Comment (lasers) [reply py3status/module.py]: I'll wait for this. :-)

File: py3status/module.py (hunk 6 above)
Comment (lasers) [reply py3status/module.py]: `../module.py|328 col 25 warning| line break after binary operator [W504]`

We have this in `tox.ini` to skip those. 
```ini
[pytest]
flake8-ignore =
# line break after binary operator (2018-Aug-22)
    W504

# invalid escape sequence (2018-Aug-22)
    W605
```

When we `black`, we can remove `W504` later so Travis will complain next time. `W605` is complaining about several `\?if=`. I'm just seeing this warning here and it didn't take much to remove the warning. I'll undo this.

Comment (tobes) [reply py3status/module.py]: but wtf is this being done in this PR for - how many times do I have to say the same shit?

Comment (lasers) [reply py3status/module.py]: Because that's how I fixed errors/warnings in first place. I forget I had this blacklisted.

File: py3status/module.py [hunk 8]
```diff
@@ -350,6 +350,35 @@ def set_module_options(self, module):
...
+                        .format(name, param)
+                    )
+
+                self.py3_module_options['block'][name] = (
+                    Py3(self)._get_color(param)
+                )
+
+                if name != 'border':
+                    continue
+                borders = ('top', 'bottom', 'left', 'right')
```
Comment (tobes) [inline py3status/module.py]: I'd just prefer the real list than building them straight after, but not a biggy.

Why tuples not lists?

File: py3status/module.py [hunk 9]
```diff
@@ -350,6 +350,35 @@ def set_module_options(self, module):
 
             self.i3s_module_options['separator_block_width'] = sep_block_width
 
+        for name in ('background', 'border'):
+            if name in mod_config:
+                param = mod_config[name]
+                if not isinstance(param, basestring):
+                    raise TypeError(
+                        'Invalid attribute `{}`, '
+                        'should be an int. Got `{}`'
```
Comment (tobes) [inline py3status/module.py]: I think you did a copy paste `string` not `int` but really should be `color`

File: py3status/module.py [hunk 10]
```diff
@@ -350,6 +350,35 @@ def set_module_options(self, module):
...
+                        .format(name, param)
+                    )
+
+                self.py3_module_options['block'][name] = (
+                    Py3(self)._get_color(param)
+                )
+
+                if name != 'border':
+                    continue
```
Comment (tobes) [inline py3status/module.py]: what's going on here?  I think you should just end the block here and unindent the following code seeing as you then screw the name variable over.

File: py3status/module.py (hunk 5 above)
Comment (tobes) [reply py3status/module.py]: is there anything special about the block stuff?  If not why not just all live happily together?

Comment (lasers) [reply py3status/module.py]: Also, if users didn't specify anything, the block is empty so I'm not wasting time looping through nothing. Can be either `background` only, `border` stuffs only, or all of them.

File: py3status/module.py (hunk 10 above)
Comment (lasers) [reply py3status/module.py]: `background` and `border` have identical code. I'm looping with two names. If users entered something... and it's background, `continue`. If it's border, then we make a quick list of borders, then start checking for inputs from users. Fine with screwing the name variable because I'm finished with `border` part (code is already added in block). Am working on getting border widths (int) at this time.

File: py3status/module.py (hunk 8 above)
Comment (tobes) [reply py3status/module.py]: please answer the question.

If you don't give me what I need how can I do a review?   Also please try to review your comments I find some very hard to follow.  Other times you can be clear so you have the skill if you are bothered, please be.

Comment (lasers) [reply py3status/module.py]: I'm serious. Is it not faster than the list? Immutable. I'm using it for loops so I don't need to manipulate the list.. That's what I meant by not needing to touch the tuples.

Comment (lasers) [reply py3status/module.py]: Maybe it does not matter because it does not apply to the line 359, but the next line down... Should be used with tuples, not list.

File: py3status/module.py (hunk 10 above)
Comment (tobes) [reply py3status/module.py]: Do you read what I write? 

I'm pretty sure it's wrong - yeah it runs but it is messy for no reason and is a maintenance nightmare.  If you are doing something clever where is the comment.

Have a fun evening I'm done :)

Comment (lasers) [reply py3status/module.py]: You always say "What's going on here?" I explained it, but I know you know what's happening here... so why are you asking me that question? Maybe no question next time so I don't have to answer your question.

That's all I did... answering that question. I did read what you said. ;-)

File: py3status/module.py (hunk 8 above)
Comment (tobes) [reply py3status/module.py]: please use a list it is clearer and no real other difference here.  If you want to save cycles there are better ways - readability wins over pretty much anything else.

File: py3status/module.py (hunk 10 above)
Comment (lasers) [reply py3status/module.py]: Done. :-) I keep tuples because it looks ugly / out of place if I list em all.

File: py3status/module.py (hunk 8 above)
Comment (lasers) [reply py3status/module.py]: Okay, I'll switch to list right now.

File: py3status/module.py (hunk 5 above)
Comment (tobes) [reply py3status/module.py]: whatever this needs a reworking at some point to make it more efficient/readable so I can live with it but make a comment explaining why this is in the code so the reasoning isn't lost

File: py3status/module.py (hunk 10 above)
Comment (tobes) [reply py3status/module.py]: ok so the code was ok except it had no comment.
github sometimes makes it hard to see the actual code

but this is much clearer so I prefer it.  Clarity is everything - make your code easy to read for idiot reviewers and your life is easier too

File: py3status/module.py (hunk 5 above)
Comment (lasers) [reply py3status/module.py]: Left a comment.

Comment (lasers) [reply py3status/module.py]: I have separated them to make it more readable. Three dict options. `py3status` dict will not have more than two options max.

File: py3status/module.py [hunk 11]
```diff
@@ -310,8 +310,12 @@ def set_module_options(self, module):
         Set universal module options to be interpreted by i3bar
         https://i3wm.org/i3status/manpage.html#_universal_module_options
         """
+        # keep i3bar module options in i3s module options.
+        # keep py3status module options in py3s module options.
+        # keep i3-gaps module options in gaps module options.
         self.i3s_module_options = {}
         self.py3_module_options = {}
+        self.gaps_module_options = {}
```
Comment (tobes) [inline py3status/module.py]: looking much better now 

`self.gaps_module_options`  I don't like this name `self.i3_gaps_module_options`?

Also can the comment be by the definition not all above everything.  I should not have to scan the code to see things - it should all be super clear.

As an aside - I think background is fine with classic `i3-bar` for some time now (border is there but has issues so not practical)

we still need the `_get_color(border)` fix and do you also want the alpha channel?

Comment (lasers) [reply py3status/module.py]: I made variable names more clear enough to not warrant a comment.

Comment (lasers) [reply py3status/module.py]: https://github.com/Airblader/i3#background--color

It should be already available with `i3bar -t`. The exception is that this wouldn't work with CSS color names and general colors unless you have a solution in mind for those particular names. We also could support `rgba()` in the future. I think users could be more likely to use color hex over `rgba()` so maybe we don't really need to add anything here.

Comment (lasers) [reply py3status/module.py]: @tobes Ping... How goes the `_get_color()` fix since it's the only thing preventing this PR from getting merged... I'm also okay with forgoing alpha channel if it's not an easy fix. Thanks.

**Comment by lasers** [comment]
@maximbaz You should try this next!!!!
```python
order += "static_string left"
order += "static_string center"
order += "static_string right"
order += "static_string block"
static_string left {
    format = 'l[\?color=good&show e]'
    format += '[\?color=lightblue&show f][\?color=red&show t]'

    border = '#EB709B'
    position = "left"
    min_length = 20

    border_top = 2
    border_bottom = 2
    border_left = 2
    border_right = 2
}
static_string center {
    format = '\?color=degraded center'
    format = 'c[\?color=degraded&show e][\?color=pink&show n]'
    format += '[\?color=orange&show t][\?color=gold&show e]r'

    border = 'orange'
    background = '#005577'
    position = "center"
    min_length = 20

    border_top = 2
    border_bottom = 2
    border_left = 2
    border_right = 2
}
static_string right {
    format = '\?color=bad right'
    format = 'r[\?color=bad&show i][\?color=lightgreen&show g]'
    format += 'h[\?color=lightblue&show t]'

    border = 'lightgreen'
    min_length = 20
    position = "right"
    border_top = 2
    border_bottom = 2
    border_left = 2
    border_right = 2
}

static_string block {
    format = '\?color=red ❤'

    min_length = 5
    background = 'gold'
    position = 'center'
    border = 'white'

    border_top = 2
    border_bottom = 2
    border_left = 2
    border_right = 2
}
```

**Comment by maximbaz** [comment]
It's fun no doubt 🙂 

I don't see how I personally will use this feature yet, but as I said in the other ticket, in principle I'm not against this PR because it enables functionality that i3bar (or i3status? I'm confused a bit) _already_ supports, currently py3status prevents using some options that what i3bar/i3status provides.

@su8, just curious, I see you upvoted this PR, could you tell us if you are going to use this feature? Maybe you already know how exactly, for which modules? I'm asking because this has been a point of discussion, whether there are some actual use-cases for this PR.

**Comment by lasers** [comment]
Users don't have to use all borders. They can set `0` for all borders except `top` or `bottom` and each module can have its own `top`/`bottom` color. The next PR (global options) will allow users to set `border_{top,bottom,right,left}` once in `py3status {}` instead of all modules then on said modules, just override it with new int value. (much fewer configs).

>...enables functionality that i3bar (or i3status? I'm confused a bit) already supports, currently py3status prevents using some options that what i3bar/i3status provides.

(i3bar). This works with `i3-gaps`. This works brokenly with `i3` too. py3status passes the options and we validate/apply the options to each composite with some exceptions. 


@su8 does not use i3wm+py3status. @su8 is a good sport. <3

**Comment by maximbaz** [comment]
`i3-gaps` is widely popular, so it's okay.

Anyhow, if you wanted my feedback, you have it now: yes for this PR (because it allows using already existing i3bar features) and no for `updates` module (because it's just a special case of `external_script`, which already provides the same functionality, plus #1439 makes `external_script` even more feature-rich).

**Comment by ultrabug** [comment]
Ok @lasers the PR is now thin and I have found a way to get rid of the get_color function as you can see in my added commit.

Could you test it one last time? It's working for me, so merge is soon to come.

**Comment by ultrabug** [comment]
@lasers hmm I can't reproduce this and it shows on a module that's not part of the project, can you repeat the problem on other modules?

**Comment by lasers** [comment]
I don't see this problem on my usual modules. The colors is working now without `pew` and `updates`. After commenting out `pew`, I can't reproduce this with `updates` alone.

**Comment by ultrabug** [comment]
Why then? :(

**Comment by lasers** [comment]
I don't have a clue. It is just `pew` so far. I uncommented several lingering module configs and can't reproduce this error anywhere else. I enabled `pew` and `updates` at bottom of the config and it worked fine now. Was it caches or something?

**Comment by lasers** [comment]
@ultrabug If you `find . -name __pycache__` in py3status and delete them, does `pew` work for you now?

**Comment by lasers** [comment]
@ultrabug Hi. I have 6 open PRs... Plz get this junk out of here today. :fire::wastebasket:

**Comment by ultrabug** [comment]
Thanks, works fine for me

**Comment by lasers** [comment]
@ultrabug With this merged, do you want me to reopen/fix up #1481?

**Comment by ultrabug** [comment]
@lasers if you make it a simple and one-thing-at-a-time PR, why not

**Comment by ultrabug** [comment]
Then just open a PR that says this clearly. TBH the previous one was messy to me.
